### PR TITLE
BTA-1674-Add-Ability-In-Our-SDK-To-Parse-And-Validate-Webhook-Headers-Ruby

### PR DESCRIPTION
**Pull request by linniethepooh**
Added headers and url to the webhook content example and made used of the `validate_webhook_request` method by validating the webhook headers before allowing them to be parsed.

https://github.com/bitpesa/bitpesa-api-specification/pull/99

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -9,6 +9,11 @@ GEM
   remote: https://rubygems.org/
   specs:
     ZenTest (4.11.1)
+    activesupport (5.2.3)
```
Comment (sztupy) [inline Gemfile.lock]: note that this gem is going to be used in the wild, so the less dependencies the better. This at the moment would for example tie this gem to Rails 5.2.

Also note as far as I know Gemfile.lock should not be commited for public gems, only the Gemfile itself

File: example/client.rb [hunk 2]
```diff
@@ -212,47 +213,63 @@ def create_and_fund_transaction_example
...
-    webhook_content = <<-JSON
-      {
-        "webhook": "85d11cf4-d4d6-46e4-ab7d-91355e80392a",
-        "event": "transaction.created",
-        "object": {
-          "input_amount": 21.59,
-          "input_currency": "GBP",
-          "id": "cd390e83-a9de-404d-a660-4da0f1342c04",
-          "metadata": {},
+    headers = <<-JSON
```
Comment (sztupy) [inline example/client.rb]: no need to have this as a JSON string - we can expect that the user will get the headers in a nice format from the library used.

Comment (sztupy) [reply example/client.rb]: Also we only need the nonce, key and secret here, not the rest

File: example/client.rb [hunk 3]
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
...
-      puts document_webhook.object.to_s
-    else
-      puts webhook
+    webhook = webhook_api.parse_response(webhook_content.body, "Webhook")
+    nonce = JSON.parse(webhook_content.headers)['Authorization-Nonce']
+    signature = JSON.parse(webhook_content.headers)['Authorization-Signature']
+    webhook_url = webhook_content.url
+    body = webhook_content.body
+
+    if webhook_api.validate_webhook_request(webhook_url, body, nonce, signature)
```
Comment (sztupy) [inline example/client.rb]: we might need to test as well that they key we got is the same key we store for the client

File: example/client.rb [hunk 4]
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
       }
     JSON
 
+    webhook_content = ResponseStruct.new(
```
Comment (sztupy) [inline example/client.rb]: This also looks a bit overcomplicated for this example. Do we need this?

File: example/client.rb [hunk 5]
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
...
-      puts payout_method_webhook.object.to_s
-    elsif webhook['event'].start_with?('sender')
-      sender_webhook = webhook_api.parse_response(webhook_content, 'SenderWebhook')
-      puts sender_webhook.object.to_s
-    elsif webhook['event'].start_with?('document')
-      document_webhook = webhook_api.parse_response(webhook_content, 'DocumentWebhook')
-      puts document_webhook.object.to_s
-    else
-      puts webhook
+    webhook = webhook_api.parse_response(webhook_content.body, "Webhook")
```
Comment (sztupy) [inline example/client.rb]: we can move this line after the first if. No need to parse the response if it doesn't even validate our security - in fact we shouldn't touch it before validating the headers for security

File: example/client.rb [hunk 6]
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
...
+    nonce = JSON.parse(webhook_content.headers)['Authorization-Nonce']
```
Comment (sztupy) [inline example/client.rb]: as above: no need for this complex setup. It's enough to have nonce = headers['Authorization-Nonce'], etc. We can expect that the headers will be in a nice format here as most client libraries will return them in a hash, and never in JSON

File: example/client.rb (hunk 3 above)
Comment (sztupy) [reply example/client.rb]: Also where was this `validate_webhook_request` method defined?

File: Gemfile.lock (hunk 1 above)
Comment (archmagos) [reply Gemfile.lock]: Adding to this, your build will fail unless you remove the lockfile. Your Gemfile specifies ZenTest 4.11.2 which is necessary to pass Travis, but the lockfile sets ZenTest to 4.11.1. It's an issue that  [this OpenApi release](https://github.com/OpenAPITools/openapi-generator/releases/tag/v4.0.0-beta3) will fix, but for now make sure the build is using v4.11.2.

File: example/client.rb (hunk 3 above)
Comment (emanuelegorga) [reply example/client.rb]: Hey Zsolt I have been pairing with Linnie on this - the method `validate_webhook_request` is contained in api-specification project. Here is the link of the PR: [API-SPECIFICATION](https://github.com/bitpesa/bitpesa-api-specification/pull/99)

File: Gemfile.lock (hunk 1 above)
Comment (linniethepooh) [reply Gemfile.lock]: thanks guys, changes made!

File: example/client.rb (hunk 3 above)
Comment (linniethepooh) [reply example/client.rb]: tested the above, and liked the other PR in the commit message.

File: example/client.rb (hunk 6 above)
Comment (linniethepooh) [reply example/client.rb]: change made

**Review by archmagos — COMMENTED** [review]
Your build is failing as you still have the lockfile committed, which is causing conflicts. If you remove it from the project and then commit that change alongside your `client.rb` file, the build should pass.

**Pull request closed without merging**